AIP-5815 -- added new Modal

reno1979 opened the pull request:
Modal to new design

Modal updates:
- optional divider for topbar and actions
- added additional tests
- added storybook controls 
- added storybook stories

Storybook: 
- react-doc-gen config added

reno1979 committed:
storybook controls for modal
added additional Modal stories
added additional Modal tests
Modal new design

reno1979 committed:
ts lint fix

reno1979 committed:
renamed var

MikeClickr commented:
@reno1979 Laat bezig ;-) Waar kan ik deze update zien? Ik test UX in de acceptatieomgeving

fortunato commented on src/common/src/Modal/ModalTopBar/ModalTopBar.tsx:
```diff
@@ -19,27 +20,28 @@ const ModalTopBar = ({ hideCloseButton = false, onCloseButton, children }: Props
-		<ModalTopBarStyle>
-			<Heading forwardedAs="div" style={{ flexGrow: 1 }}>
+		<ModalTopBarStyle data-testid="modal-top-bar" hideDivider={hideDivider}>
+			<div className="truncated" data-testid="modal-top-bar-children">
```
Did you consider that using a class name like this does not ultimately encapsulate/protect your styles from being overridden by another component's styles?
![image](https://user-images.githubusercontent.com/1095870/179161526-1b6633a3-170b-48ae-9ca3-deb5f02f1201.png)
One could argue that creating a new styled component for this div would be a more styled-components way of doing things. I.e.:
```typescript
export const Truncated = styled.div`
    // ...
`
```

reno1979 replied:
You are right, I'm used to custom elements with a shadowroot.

fortunato commented on src/common/src/Modal/Modal.stories.tsx:
```diff
@@ -1,30 +1,30 @@
 
-const Template: Story<React.ComponentProps<typeof Modal>> = (args) => {
-	const [isModalVisible, setModalVisibility] = React.useState<boolean>(false);
+const Template: ComponentStory<typeof Modal> = (args) => {
```
Cool!

fortunato commented on src/common/src/Modal/Modal.tsx:
```diff
@@ -15,14 +15,16 @@ type Props = {
+// https://github.com/storybookjs/storybook/issues/13408
+export const Modal: IModal = ({ id, children, classnames, onClose, size = 'md', disablePortal, ...rest }: Props) => {
 	const handleClose = () => {
+		console.log('handleClose');
```
This can be removed?

fortunato commented on src/common/src/Modal/Modal.tsx:
```diff
@@ -15,14 +15,16 @@ type Props = {
 }
-
-const Modal: IModal = ({ id, children, classnames, onClose, size = 'md', disablePortal, ...rest }: Props) => {
+// https://github.com/storybookjs/storybook/issues/13408
```
Oh wow. +1 for not using default exports :thinking: 
(https://basarat.gitbook.io/typescript/main-1/defaultisbad)

fortunato commented on src/common/src/Modal/ModalActions/ModalActions.test.tsx:
```diff
@@ -0,0 +1,43 @@
+	 *  Can't use getBoundClientRect because jsdom only returns the correct keys, not the actual values.
+	 *  See: https://github.com/jsdom/jsdom/issues/653
+	 */
+	it.skip('should render with expected styles', () => {
```
Would this perhaps be better left as a `it.todo()`? Or does a comment without a test case suffice?

fortunato commented on src/common/src/Modal/ModalActions/ModalActionsStyles.ts:
```diff
@@ -1,28 +1,40 @@
+	${Divider} {
+		grid-column: 1 / span 2;
+		grid-row: 1;
+		align-self: flex-start;
```
I presume this line can be removed after switching from flexbox to grid?

fortunato commented on src/common/src/Modal/ModalTopBar/ModalTopBar.test.tsx:
```diff
@@ -0,0 +1,96 @@
+import React from 'react';
```
Looks pretty thorough, nice.
Did you consider testing for `.toBeVisible()` instead of just `.toBeDefined()`? I personally prefer`.toBeInTheDocument()` as it might add some semantic value to the reader over toBeDefined.
Overall it's a bit heavy on the use of data-testid selectors one could argue: https://derekndavis.com/posts/getbytestid-overused-react-testing-library
Thoughts?

reno1979 replied:
I agree that it is a bit heavy on the testid attributes, but the use of styled components generates random classes and selectors like `nth-child` or header > div etc will cause a lot of required selector maintenance. Adding an additional class to an element can cause styling collisions (because no shadowroot). Perhaps it takes a bit of patience to accept styled-components, a lot of patience.

fortunato commented on src/common/src/Modal/ModalTopBar/ModalTopBar.test.tsx:
```diff
@@ -0,0 +1,96 @@
+		const style = window.getComputedStyle(divider);
+		expect(style.backgroundColor).not.toBe('transparent');
+	});
+	/**
```
Missing newline :grin:

fortunato commented on src/common/src/Modal/ModalTopBar/ModalTopBar.test.tsx:
```diff
@@ -0,0 +1,96 @@
+	 */
+	it('should render with expected styles', () => {
+		const { container } = render(<ModalTopBar>Foo</ModalTopBar>);
+		// const header = getByTestId(container, 'modal-top-bar');
```
Can be removed?

reno1979 replied:
I would like to keep this inline, because of the mentioned jsdom issue not being able to test on a actual size

fortunato replied:
Ok, fair enough

fortunato commented on src/common/src/Modal/ModalTopBar/ModalTopBar.test.tsx:
```diff
@@ -0,0 +1,96 @@
+	 *  Can't use getBoundClientRect because jsdom only returns the correct keys, not the actual values.
+	 *  See: https://github.com/jsdom/jsdom/issues/653
+	 */
+	it.skip('should truncate long content, test by verify height', () => {
```
* "verifying" :stuck_out_tongue:

fortunato reviewed (commented):
Nice work overall. Looks like you're getting the hang of React & friends :+1:

reno1979 commented on src/utils/storybook.ts:
```diff
@@ -0,0 +1,14 @@
+export const DISABLED_CONTROL = { description: 'Disabled control for this story', control: false };
+
+export function generateDisabledControls(propertyNames: string[] = []): { [key: string]: typeof DISABLED_CONTROL } {
+	if (!propertyNames.length) {
```
nice one!

reno1979 committed:
MR updates

reno1979 committed:
MR test fixes

reno1979 committed:
bugfix overflow

reno1979 committed:
removed skipped test

reno1979 commented:
@MikeClickr Eventueel kan ik een statische versie van storybook maken

reno1979 commented:
van @MikeClickr :: 

Het ziet er goed uit, heb wel paar vragen, reviews:

1. Een modal zonder close button mag van mij weg. Sluit knop is nodig (A11Y) om met toetsenbord modal te kunnen sluiten.
2. Kun je sluiten modal ook met [Esc] key en klik naast modal werkend maken?
3. De lijn in header en footer, is true/false in te stellen? Default = false, want dat is volgens design system. Alleen bij modals zoals bulkmetadatering bij DMS is een scroll nodig en daarmee de lijnen.
4. Als je 200% vergroot, worden header en footer ook hoger, zo eet je de content op. Zou goed zijn om header en footer te maximeren.

Kun je hier iets mee?

---- antwoord van mij ----

1. Dat gaat wel lukken
2. Die viel me ook op, dat gaat niet 123 lukken denk ik omdat het modal gebaseerd is op Amsterdam Styled Components en daar werkt het ook niet
3. De lijnen zijn nu reeds optioneel te zetten (hideDivider), default niet zichtbaar is geen probleem dat pas ik aan
4. Daar kan ik weinig mee, als je vergroot (zoomt) dan zoomt alles. De header is idd er groot en vraagt meer aandacht dan de content, waar wel wat mee gedaan kan worden is ontwerpen voor kleinere schermen. De modal is nu gemaakt voor desktop gebruik, terwijl we imo naar mobile first moeten

reno1979 committed:
modal dividers hidden by default
modal close button visible by default

reno1979 committed:
Merge branch 'main' into story/aip5815

reno1979 committed:
storybook update
added npmrc
ModalContent styling fix first and last element
ModalTopBar added bottom margin when hideDivider is false

reno1979 committed:
lint fixes

reno1979 commented:
van @MikeClickr 

Ik zie in de modals nog wel paar issues:
- Heading te klein font
- Te veel marge tussen heading en body
Zie UX: https://www.figma.com/file/4a6C5fMGJ8dd7ZQgdLVcnf/AIP_gemeenteAmsterdam?node-id=683%3A15050

Deze punten zijn inmiddels ook verwerkt

reno1979 committed:
added gap to modal actions

reno1979 committed:
documentTable disabled all test that change state, possible memory leak

reno1979 committed:
added todo

reno1979 committed:
silent ci:test

reno1979 commented:
closing this, re-created the branch to work with new updates
